src/serde/deserialize_utils.rs

```diff
@@ -0,0 +1,359 @@
+use crate::serde::deserialize_program::ValueAddress;
+use crate::types::instruction::Register;
+use num_bigint::{BigInt, ParseBigIntError};
+use std::fmt;
+use std::num::{IntErrorKind, ParseIntError};
+use std::ops::Rem;
+
+#[derive(Debug, PartialEq)]
+pub enum ReferenceParseError {
+    IntError(ParseIntError),
+    BigIntError(ParseBigIntError),
+    InvalidStringError(String),
+}
+
+impl fmt::Display for ReferenceParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ReferenceParseError::IntError(error) => {
+                write!(f, "Int parsing error: ")?;
+                error.fmt(f)
+            }
+            ReferenceParseError::BigIntError(error) => {
+                write!(f, "BigInt parsing error: ")?;
+                error.fmt(f)
+            }
+            ReferenceParseError::InvalidStringError(error) => {
+                write!(f, "Invalid reference string error: ")?;
+                error.fmt(f)
+            }
+        }
+    }
+}
+
+// Checks if the hex string has an odd length.
+// If that is the case, prepends '0' to it.
+pub fn maybe_add_padding(mut hex: String) -> String {
+    if hex.len().rem(2) != 0 {
+        hex.insert(0, '0');
+        return hex;
+    }
+    hex
+}
+
+pub fn parse_dereference(value: &str) -> Result<ValueAddress, ReferenceParseError> {
+    let splitted: Vec<&str> = value.split(" + ").collect();
+
+    match splitted.len() {
+        1 => parse_dereference_no_offsets(splitted),
+        2 => parse_dereference_with_one_offset(splitted),
+        3 => parse_dereference_with_two_offsets(splitted),
+        _ => Err(ReferenceParseError::InvalidStringError(String::from(value))),
+    }
+}
+// parse string values of format `[cast(reg, *felt)]`
+fn parse_dereference_no_offsets(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let str_tmp: Vec<&str> = splitted_value_str[0].split(',').collect();
+
+    let register = match str_tmp[0].split('(').collect::<Vec<_>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
+
+    Ok(ValueAddress {
+        register,
+        offset1: 0,
+        offset2: 0,
+        immediate: None,
+        dereference: true,
+    })
+}
+
+// parse string values of format `[cast(reg + offset1, *felt)]`
+fn parse_dereference_with_one_offset(
+    splitted_value_str: Vec<&str>,
+) -> Result<ValueAddress, ReferenceParseError> {
+    let register = match splitted_value_str[0].split('(').collect::<Vec<&str>>()[1] {
+        "ap" => Some(Register::AP),
+        "fp" => Some(Register::FP),
+        _ => None,
+    };
```

Comment:
```suggestion
    let mut deref = parse_dereference_no_offsets(splitted_value_str);
```

Then you can overwrite `deref.offset1` right before returning.